SudokuSolver/SudokuSolver/SudokuSolver.cs

```diff
@@ -0,0 +1,137 @@
+        private int _rowSqrt;
+        private int _colSqrt;
+
+        private List<EmptySudokuCube> _emptyCubes;
+
+        private void InitEmptyCubes()
+        {
+            _emptyCubes = new List<EmptySudokuCube>();
+
+            for (int i = 0; i < _maxRows; i++)
+            {
+                for (int j = 0; j < _maxCols; j++)
+                {
+                    if (_grid[i, j] == 0)
+                    {
+                        var eSC = new EmptySudokuCube(i, j);
+                        // Check all legal nums and add them to cube's possible values
+                        eSC.PossibleValues.AddRange(
+                            Enumerable.Range(1, _maxRows).Where(num => IsLegal(i, j, num)).ToList()
+                        );
+                        _emptyCubes.Add(eSC);
+                    }
+                }
+            }
+        }
+
+        public SudokuSolver(int [, ] grid)
+        {
+            _grid = grid;
+            _maxRows = grid.GetLength(0);
+            _maxCols = grid.GetLength(1);
+
+            if (_maxCols != _maxRows) throw new ArgumentException("Matrix should have the same number of rows and columns.");
+
+            _rowSqrt = (int)Math.Sqrt(_maxRows);
+            _colSqrt = (int)Math.Sqrt(_maxCols);
+
+            InitEmptyCubes();
+        }
+
+        public bool Solve()
+        {
+            if (_emptyCubes.Count() == 0) return true;
+
+            // Take empty cube with the least possibilities values 
+            var eSC = _emptyCubes.OrderByDescending(eSC => eSC.PossibleValues.Count()).Last();
+
+            foreach (var value in eSC.PossibleValues)
+            {
+                if (IsLegal(eSC.Row, eSC.Col, value))
+                {
+                    _grid[eSC.Row, eSC.Col] = value;
+                    _emptyCubes.Remove(eSC);
+
+                    // Use recursion to check other empty sudoku cubes
+                    if (Solve()) return true;
+
+                    // Undo and check another possible value 
+                    _emptyCubes.Add(eSC);
+                    _grid[eSC.Row, eSC.Col] = 0;
+                }
+            }
+            return false;
+        }
+
+        private bool IsLegal(int rowIdx, int colIdx, int num)
+        {
+            var row = _grid.GetRow(rowIdx);
+            var col = _grid.GetColumn(colIdx);
+
+            // Check if we find the same num in the similar row
+            if (row.Count(n => n == num) > 0) return false;
+
+            // Check if we find the same num in the similar column
+            if (col.Count(n => n == num) > 0) return false;
+
+            // Check if we find the same num in the particular 3*3 matrix
+            int startRow = rowIdx - rowIdx % _rowSqrt;
+            int startCol = colIdx - colIdx % _colSqrt;
+
+            for (int i = startRow; i < startRow + _rowSqrt; i++)
+                for (int j = startCol; j < startCol + _colSqrt; j++)
+                    if (_grid[i, j] == num) return false;
```

Comment:
ten komentarz kłamie. To niekoniecznie jest 3x3 matrix.
Też można to wydzielić do funkcji z nazwą typu "isNumberAlreadyInSquare" zamiast komentarza